runtime/src/main/cpp/Memory.cpp

```diff
@@ -931,6 +931,9 @@ void freeAggregatingFrozenContainer(ContainerHeader* container) {
 void runDeallocationHooks(ContainerHeader* container) {
   ObjHeader* obj = reinterpret_cast<ObjHeader*>(container + 1);
   for (int index = 0; index < container->objectCount(); index++) {
+    if (obj->type_info() == theSharedRefTypeInfo) {
```

Comment:
Does this have any effect on the performance for other classes?

Reply:
Good thing you asked. I didn't think it would have a measurable impact since it runs during deallocation, it's a simple compare and jump, and `obj->type_info()` is accessed below regardless.

And then I ran `ring` benchmarks:
```
Performance details
=================
Benchmark                                         First score              Second score             Percent                  Ratio                    
------------------------------------------------------------------------------------------------------------------------------------------------------
Geometric mean                                    221.4273 ± 0.8246        220.2083 ± 0.8864        0.00 ± 0.00 %            1.01 ± 0.01              
------------------------------------------------------------------------------------------------------------------------------------------------------
Ring::Euler.problem9                              852.9136 ± 1.6820        574.3120 ± 1.8881        32.31 ± 0.44 %           1.49 ± 0.01              
Ring::Calls.interfaceMethodBimorphic              566.7952 ± 5.0681        394.2084 ± 0.8376        29.67 ± 1.25 %           1.44 ± 0.02              
Ring::Calls.openMethodTrimorphic                  1982.0209 ± 6.1022       1481.8918 ± 10.8697      24.45 ± 1.10 %           1.34 ± 0.01              
Ring::IntArray.countFilteredSomeLocal             22.6207 ± 0.0427         17.5005 ± 0.0401         22.31 ± 0.36 %           1.29 ± 0.01              
Ring::CompanionObject.invokeRegularFunction       0.0056 ± 0.0000          0.0046 ± 0.0000          16.12 ± 0.52 %           1.20 ± 0.01              
Ring::IntArray.countFilteredSomeManual            20.4322 ± 0.0561         17.5310 ± 0.0482         13.73 ± 0.47 %           1.17 ± 0.01              
Ring::String.stringBuilderConcat                  287.8454 ± 1.9528        255.7878 ± 2.8703        9.53 ± 2.01 %            1.13 ± 0.02              
Ring::String.stringBuilderConcatNullable          288.5585 ± 2.8659        256.7720 ± 2.3562        9.30 ± 1.78 %            1.12 ± 0.02              
Ring::ParameterNotNull.invokeTwoArgsWithNullCheck 0.0090 ± 0.0000          0.0081 ± 0.0000          9.08 ± 0.65 %            1.11 ± 0.01              
Ring::Euler.problem2                              6.8504 ± 0.0900          6.1866 ± 0.0220          8.16 ± 2.38 %            1.11 ± 0.02              
Ring::Euler.problem8                              339.8889 ± 1.4641        309.6184 ± 2.6217        7.74 ± 1.55 %            1.10 ± 0.01              
Ring::DefaultArgument.testFourOfFour              0.0051 ± 0.0000          0.0047 ± 0.0000          7.23 ± 1.63 %            1.09 ± 0.01              
Ring::ClassArray.countFilteredLocal               2497.6665 ± 20.3156      2293.5391 ± 6.3643       7.16 ± 1.50 %            1.09 ± 0.01              
Ring::Calls.interfaceMethodHexamorphic            2455.9505 ± 17.4846      2276.1167 ± 5.1275       6.45 ± 1.32 %            1.08 ± 0.01              
Ring::IntList.filterAndCount                      2617.9088 ± 10.9396      2481.8600 ± 17.6619      4.12 ± 1.35 %            1.05 ± 0.01              
Ring::ForLoops.arrayLoop                          7.5842 ± 0.0450          7.1792 ± 0.0610          3.96 ± 1.62 %            1.06 ± 0.02              
Ring::IntList.filterManual                        2595.6617 ± 2.9685       2488.8430 ± 4.4040       3.84 ± 0.34 %            1.04 ± 0.00              
Ring::OctoTest                                    3392.8933 ± 24.0985      3235.5541 ± 7.6086       3.73 ± 1.36 %            1.05 ± 0.01              
Ring::AbstractMethod.sortStringsWithComparator    227.7961 ± 0.9786        220.8034 ± 0.9333        2.24 ± 0.84 %            1.03 ± 0.01              
Ring::IntArray.copy                               1724.3959 ± 2.9719       1690.3470 ± 4.5611       1.54 ± 0.53 %            1.02 ± 0.00              
Ring::Calls.openMethodBimorphic                   394.3895 ± 1.7671        561.6845 ± 0.8763        -29.36 ± 0.63 %          0.70 ± 0.00              
Ring::ForLoops.charArrayLoop                      3.3796 ± 0.0171          4.6797 ± 0.0165          -27.16 ± 0.73 %          0.72 ± 0.01              
Ring::Calls.interfaceMethodTrimorphic             1498.6024 ± 4.4771       1981.0271 ± 5.7760       -23.90 ± 0.45 %          0.76 ± 0.00              
Ring::ParameterNotNull.invokeOneArgWithoutNullCheck0.0077 ± 0.0000          0.0099 ± 0.0000          -21.84 ± 0.59 %          0.78 ± 0.01              
Ring::Elvis.testElvis                             72.2298 ± 0.0781         85.2601 ± 0.2739         -14.92 ± 0.54 %          0.85 ± 0.00              
Ring::LocalObjects.localArray                     0.0856 ± 0.0006          0.1016 ± 0.0010          -14.35 ± 1.68 %          0.84 ± 0.01              
Ring::IntArray.countFilteredSome                  17.5423 ± 0.0493         20.3742 ± 0.0242         -13.55 ± 0.48 %          0.86 ± 0.00              
Ring::ClassStream.filterAndMapManual              2528.4038 ± 2.9074       2856.4001 ± 20.7141      -10.73 ± 1.29 %          0.89 ± 0.01              
Ring::IntBaseline.allocateArrayAndFill            26.7258 ± 0.1895         29.6828 ± 0.2427         -8.58 ± 1.48 %           0.90 ± 0.01              
Ring::WithIndicies.withIndiciesManual             2449.8847 ± 3.1263       2685.0674 ± 18.4686      -8.01 ± 1.26 %           0.91 ± 0.01              
Ring::ClassBaseline.consumeField                  58.4672 ± 0.2731         63.9413 ± 0.5942         -7.27 ± 1.71 %           0.91 ± 0.01              
Ring::Switch.testConstSwitch                      187.1804 ± 0.5415        202.5431 ± 0.6270        -7.03 ± 0.57 %           0.92 ± 0.01              
Ring::ParameterNotNull.invokeTwoArgsWithoutNullCheck0.0077 ± 0.0000          0.0082 ± 0.0001          -5.00 ± 2.14 %           0.94 ± 0.01              
Ring::Loop.arrayLoop                              66.5391 ± 0.5565         70.6445 ± 0.4824         -4.37 ± 1.59 %           0.94 ± 0.01              
Ring::Loop.arrayListLoop                          155.6977 ± 1.3134        164.9411 ± 0.9808        -4.24 ± 1.60 %           0.94 ± 0.01              
Ring::Loop.arrayIndexLoop                         96.0024 ± 0.4738         102.1985 ± 1.4875        -4.20 ± 2.75 %           0.94 ± 0.02              
Ring::Loop.arrayForeachLoop                       66.8288 ± 0.5850         70.4282 ± 0.4796         -3.62 ± 1.67 %           0.95 ± 0.01              
Ring::Loop.arrayListForeachLoop                   156.5447 ± 0.8745        164.1294 ± 1.3592        -3.29 ± 1.59 %           0.95 ± 0.01              
Ring::Switch.testVarSwitch                        190.1685 ± 0.5312        195.6584 ± 0.7369        -2.17 ± 0.73 %           0.97 ± 0.01              
Ring.compileTime                                  20548585.3580 ± 0.0000   20841423.6600 ± 0.0000   -1.41 ± 0.00 %           0.99 ± 0.00 
```
Some huge regressions and some huge improvements. This is all on my local machine, of course, but I reran the suite several times, and the results are stable. I'll be looking into it.

Reply:
This is caused by `runDeallocationHooks` stopping to be automatically inlined and `DisposeSharedRef` being automatically inlined and also contributing with exception handlers around it. Put appropriate annotations to restore previous behavior. It's worth investigating whether we need to put more of `noexcept` annotations around and then revisiting this manual control of inlining.